diff --git a/certs.sh b/certs.sh
index f7c19f5..4b5eccc 100755
--- a/certs.sh
+++ b/certs.sh
@@ -63,6 +63,6 @@ cp "$CERTDIR"/ca_"$CHAINLEN".key "$CERTDIR"/priv.key
 # combine certificates
 for i in $(seq "$CHAINLEN" -1 1); do
   cat "$CERTDIR"/cert_"$i".pem >>"$CERTDIR"/cert.pem
-  rm "$CERTDIR"/cert_"$i".pem "$CERTDIR"/ca_"$i".key
+  rm -f "$CERTDIR"/cert_"$i".pem "$CERTDIR"/ca_"$i".key
 done
 rm -f "$CERTDIR"/*.srl "$CERTDIR"/ca_0.key "$CERTDIR"/cert.csr
diff --git a/docker-compose.yml b/docker-compose.yml
index 93664f0..f0fa93e 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -2,7 +2,7 @@ version: "2.4"

 services:
   sim:
     image: martenseemann/quic-network-simulator
     container_name: sim
     hostname: sim
     stdin_open: true
@@ -41,6 +41,7 @@ services:
       - SSLKEYLOGFILE=/logs/keys.log
       - QLOGDIR=/logs/qlog/
       - TESTCASE=$TESTCASE_SERVER
+      - TEST_TYPE=$TEST_TYPE
     depends_on:
       - sim
     cap_add: 
@@ -68,6 +69,7 @@ services:
       - SSLKEYLOGFILE=/logs/keys.log
       - QLOGDIR=/logs/qlog/
       - TESTCASE=$TESTCASE_CLIENT
+      - TEST_TYPE=$TEST_TYPE
       - REQUESTS=$REQUESTS
     depends_on:
       - sim
diff --git a/implementations.json b/implementations.json
index 09e2fdd..72513c8 100644
--- a/implementations.json
+++ b/implementations.json
@@ -74,8 +74,13 @@
     "url": "https://github.com/quinn-rs/quinn",
     "role": "both"
   },
+  "s2n-quic-rustls": {
+    "image": "aws/s2n-quic-qns-rustls:latest",
+    "url": "https://github.com/aws/s2n-quic",
+    "role": "both"
+  },
   "s2n-quic": {
-    "image": "ghcr.io/aws/s2n-quic/s2n-quic-qns:latest",
+    "image": "aws/s2n-quic-qns:latest",
     "url": "https://github.com/aws/s2n-quic",
     "role": "both"
   },
diff --git a/interop.py b/interop.py
index d6b27ce..ac7a7b7 100644
--- a/interop.py
+++ b/interop.py
@@ -126,6 +126,7 @@ class InteropRunner:
         cmd = (
             "CERTS=" + certs_dir.name + " "
             "TESTCASE_CLIENT=" + random_string(6) + " "
+            "TEST_TYPE=TEST "
             "SERVER_LOGS=/dev/null "
             "CLIENT_LOGS=" + client_log_dir.name + " "
             "WWW=" + www_dir.name + " "
@@ -153,6 +154,7 @@ class InteropRunner:
         cmd = (
             "CERTS=" + certs_dir.name + " "
             "TESTCASE_SERVER=" + random_string(6) + " "
+            "TEST_TYPE=TEST "
             "SERVER_LOGS=" + server_log_dir.name + " "
             "CLIENT_LOGS=/dev/null "
             "WWW=" + www_dir.name + " "
@@ -408,6 +410,7 @@ class InteropRunner:
             "CERTS=" + testcase.certs_dir() + " "
             "TESTCASE_SERVER=" + testcase.testname(Perspective.SERVER) + " "
             "TESTCASE_CLIENT=" + testcase.testname(Perspective.CLIENT) + " "
+            "TEST_TYPE=" + testcase.test_type() + " "
             "WWW=" + testcase.www_dir() + " "
             "DOWNLOADS=" + testcase.download_dir() + " "
             "SERVER_LOGS=" + server_log_dir.name + " "
@@ -524,9 +527,14 @@ class InteropRunner:
         logging.debug(values)
         res = MeasurementResult()
         res.result = TestResult.SUCCEEDED
-        res.details = "{:.0f} (± {:.0f}) {}".format(
-            statistics.mean(values), statistics.stdev(values), test.unit()
-        )
+        if len(values) == 1:
+            res.details = "{:.0f} {}".format(
+                values[0], test.unit()
+            )
+        else:
+            res.details = "{:.0f} (± {:.0f}) {}".format(
+                statistics.mean(values), statistics.stdev(values), test.unit()
+            )
         return res

     def run(self):
@@ -541,23 +549,26 @@ class InteropRunner:
                 client,
                 self._implementations[client]["image"],
             )
-            if not (
-                self._check_impl_is_compliant(server)
-                and self._check_impl_is_compliant(client)
-            ):
-                logging.info("Not compliant, skipping")
-                continue

+            transfer_succeeded = True
             # run the test cases
             for testcase in self._tests:
                 status = self._run_testcase(server, client, testcase)
                 self.test_results[server][client][testcase] = status
                 if status == TestResult.FAILED:
                     nr_failed += 1
+                    if testcase == testcases.TestCaseTransfer:
+                        transfer_succeeded = False

             # run the measurements
             for measurement in self._measurements:
-                res = self._run_measurement(server, client, measurement)
+                if transfer_succeeded:
+                    res = self._run_measurement(server, client, measurement)
+                else:
+                    logging.debug("Skipping measurements as Transfer testcase was unsuccessful")
+                    res = MeasurementResult()
+                    res.result = TestResult.UNSUPPORTED
+                    res.details = "Skipping measurements as Transfer testcase was unsuccessful"
                 self.measurement_results[server][client][measurement] = res

         self._postprocess_results()
diff --git a/pull.py b/pull.py
index c2d6d1f..844bbd5 100644
--- a/pull.py
+++ b/pull.py
@@ -5,7 +5,7 @@ import sys
 from implementations import IMPLEMENTATIONS
 
 print("Pulling the simulator...")
 os.system("docker pull martenseemann/quic-network-simulator")
 
 print("\nPulling the iperf endpoint...")
 os.system("docker pull martenseemann/quic-interop-iperf-endpoint")
diff --git a/requirements.txt b/requirements.txt
index 131cf94..ce5960d 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,4 @@
 pycryptodome
 termcolor
 prettytable
-pyshark
\ No newline at end of file
+pyshark==0.5.2
\ No newline at end of file
diff --git a/run.py b/run.py
index 4564681..3a28eec 100755
--- a/run.py
+++ b/run.py
@@ -165,4 +165,4 @@ def main():
 
 
 if __name__ == "__main__":
-    sys.exit(main())
+    main()
diff --git a/testcases.py b/testcases.py
index fc69e11..7dafb07 100644
--- a/testcases.py
+++ b/testcases.py
@@ -98,6 +98,10 @@ class TestCase(abc.ABC):
         """The name of testcase presented to the endpoint Docker images"""
         return self.name()

+    @staticmethod
+    def test_type() -> str:
+        return "TEST"
+
     @staticmethod
     def scenario() -> str:
         """Scenario for the ns3 simulator"""
@@ -1239,10 +1243,14 @@ class TestCasePortRebinding(TestCaseTransfer):
         ]
         return self._files

+    @staticmethod
+    def test_type() -> str:
+        return "TEST"
+
     @staticmethod
     def scenario() -> str:
         """Scenario for the ns3 simulator"""
-        return "rebind --delay=15ms --bandwidth=10Mbps --queue=25 --first-rebind=1s --rebind-freq=5s"
+        return "rebind --delay=15ms --bandwidth=10Mbps --queue=25 --first-rebind=2s --rebind-freq=5s"

     def check(self) -> TestResult:
         super().check()
@@ -1265,56 +1273,26 @@ class TestCasePortRebinding(TestCaseTransfer):
             logging.info("Server saw only a single client port in use; test broken?")
             return TestResult.FAILED

-        last = None
-        num_migrations = 0
-        for p in tr_server:
-            cur = (
-                (
-                    getattr(p["ipv6"], "dst")
-                    if "IPV6" in str(p.layers)
-                    else getattr(p["ip"], "dst")
-                ),
-                int(getattr(p["udp"], "dstport")),
-            )
-            if last is None:
-                last = cur
-                continue
-
-            if last != cur:
-                last = cur
-                num_migrations += 1
-                # packet to different IP/port, should have a PATH_CHALLENGE frame
-                if hasattr(p["quic"], "path_challenge.data") is False:
-                    logging.info(
-                        "First server packet to new client destination %s did not contain a PATH_CHALLENGE frame",
-                        cur,
-                    )
-                    logging.info(p["quic"])
-                    return TestResult.FAILED
-
-        tr_client = self._client_trace()._get_packets(
-            self._client_trace()._get_direction_filter(Direction.FROM_CLIENT) + " quic"
+        tr_server2client = self._client_trace()._get_packets(
+            self._client_trace()._get_direction_filter(Direction.FROM_SERVER) + " quic"
         )

         challenges = list(
             set(
                 getattr(p["quic"], "path_challenge.data")
-                for p in tr_server
+                for p in tr_server2client
                 if hasattr(p["quic"], "path_challenge.data")
             )
         )
-        if len(challenges) < num_migrations:
-            logging.info(
-                "Saw %d migrations, but only %d unique PATH_CHALLENGE frames",
-                len(challenges),
-                num_migrations,
-            )
-            return TestResult.FAILED
+
+        tr_client2server = self._client_trace()._get_packets(
+            self._client_trace()._get_direction_filter(Direction.FROM_CLIENT) + " quic"
+        )

         responses = list(
             set(
                 getattr(p["quic"], "path_response.data")
-                for p in tr_client
+                for p in tr_client2server
                 if hasattr(p["quic"], "path_response.data")
             )
         )
@@ -1332,6 +1310,10 @@ class TestCaseAddressRebinding(TestCasePortRebinding):
     def name():
         return "rebind-addr"

+    @staticmethod
+    def test_type() -> str:
+        return "MEASUREMENT"
+
     @staticmethod
     def abbreviation():
         return "BA"
@@ -1615,7 +1597,7 @@ class MeasurementGoodput(Measurement):

     @staticmethod
     def repetitions() -> int:
-        return 5
+        return 1

     def get_paths(self):
         self._files = [self._generate_random_file(self.FILESIZE)]
@@ -1687,7 +1669,7 @@ TESTCASES = [
     TestCaseMultiplexing,
     TestCaseRetry,
     TestCaseResumption,
-    TestCaseZeroRTT,
+    # TestCaseZeroRTT,
     TestCaseHTTP3,
     TestCaseBlackhole,
     TestCaseKeyUpdate,
@@ -1698,12 +1680,11 @@ TESTCASES = [
     TestCaseHandshakeCorruption,
     TestCaseTransferCorruption,
     TestCaseIPv6,
-    TestCaseV2,
-    # The next three tests are disabled due to Wireshark not being able
-    # to decrypt packets sent on the new path.
-    # TestCasePortRebinding,
-    # TestCaseAddressRebinding,
-    # TestCaseConnectionMigration,
+    # s2n-quic doesn't currently support v2
+    # TestCaseV2,
+    TestCasePortRebinding,
+    TestCaseAddressRebinding,
+    TestCaseConnectionMigration,
 ]
 
 MEASUREMENTS = [
