diff --git a/certs.sh b/certs.sh
index b26b2f8..b7652af 100755
--- a/certs.sh
+++ b/certs.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 set -e
 
diff --git a/docker-compose.yml b/docker-compose.yml
index c203219..83896f5 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -40,6 +40,7 @@ services:
       - SSLKEYLOGFILE=/logs/keys.log
       - QLOGDIR=/logs/qlog/
       - TESTCASE=$TESTCASE_SERVER
+      - TEST_TYPE=$TEST_TYPE
     depends_on:
       - sim
     cap_add: 
@@ -67,6 +68,7 @@ services:
       - SSLKEYLOGFILE=/logs/keys.log
       - QLOGDIR=/logs/qlog/
       - TESTCASE=$TESTCASE_CLIENT
+      - TEST_TYPE=$TEST_TYPE
       - REQUESTS=$REQUESTS
     depends_on:
       - sim
@@ -79,6 +81,7 @@ services:
         ipv4_address: 193.167.0.100
         ipv6_address: fd00:cafe:cafe:0::100
     extra_hosts:
+      - "server:193.167.100.100"
       - "server4:193.167.100.100"
       - "server6:fd00:cafe:cafe:100::100"
       - "server46:193.167.100.100"
@@ -124,6 +127,7 @@ services:
         ipv4_address: 193.167.0.90
         ipv6_address: fd00:cafe:cafe:0::90
     extra_hosts:
+      - "server:193.167.100.100"
       - "server4:193.167.100.110"
       - "server6:fd00:cafe:cafe:100::110"
       - "server46:193.167.100.110"
diff --git a/implementations.json b/implementations.json
index f621800..06f3595 100644
--- a/implementations.json
+++ b/implementations.json
@@ -1,4 +1,9 @@
 {
+  "s2n-quic": {
+    "image": "awslabs/s2n-quic-qns:latest",
+    "url": "https://github.com/awslabs/s2n-quic",
+    "role": "both"
+  },
   "quic-go": {
     "image": "martenseemann/quic-go-interop:latest",
     "url": "https://github.com/lucas-clemente/quic-go",
diff --git a/interop.py b/interop.py
index 6ad191c..982cb57 100644
--- a/interop.py
+++ b/interop.py
@@ -124,6 +124,7 @@ class InteropRunner:
         cmd = (
             "CERTS=" + certs_dir.name + " "
             "TESTCASE_CLIENT=" + random_string(6) + " "
+            "TEST_TYPE=TEST" + " "
             "SERVER_LOGS=/dev/null "
             "CLIENT_LOGS=" + client_log_dir.name + " "
             "WWW=" + www_dir.name + " "
@@ -148,6 +149,7 @@ class InteropRunner:
         cmd = (
             "CERTS=" + certs_dir.name + " "
             "TESTCASE_SERVER=" + random_string(6) + " "
+            "TEST_TYPE=TEST" + " "
             "SERVER_LOGS=" + server_log_dir.name + " "
             "CLIENT_LOGS=/dev/null "
             "WWW=" + www_dir.name + " "
@@ -344,6 +346,7 @@ class InteropRunner:
             "CERTS=" + testcase.certs_dir() + " "
             "TESTCASE_SERVER=" + testcase.testname(Perspective.SERVER) + " "
             "TESTCASE_CLIENT=" + testcase.testname(Perspective.CLIENT) + " "
+            "TEST_TYPE=" + testcase.test_type() + " "
             "WWW=" + testcase.www_dir() + " "
             "DOWNLOADS=" + testcase.download_dir() + " "
             "SERVER_LOGS=" + server_log_dir.name + " "
diff --git a/testcases.py b/testcases.py
index 4854842..d7fb471 100644
--- a/testcases.py
+++ b/testcases.py
@@ -90,6 +90,10 @@ class TestCase(abc.ABC):
         """ The name of testcase presented to the endpoint Docker images"""
         return self.name()
 
+    @staticmethod
+    def test_type() -> str:
+        return "TEST"
+
     @staticmethod
     def scenario() -> str:
         """ Scenario for the ns3 simulator """
@@ -183,9 +187,6 @@ class TestCase(abc.ABC):
         if len(versions) != 1:
             logging.info("Expected exactly one version. Got %s", versions)
             return False
-        if QUIC_VERSION not in versions:
-            logging.info("Wrong version. Expected %s, got %s", QUIC_VERSION, versions)
-            return False
 
         if len(self._files) == 0:
             raise Exception("No test files generated.")
@@ -1445,6 +1446,10 @@ class MeasurementGoodput(Measurement):
     def testname(p: Perspective):
         return "transfer"
 
+    @staticmethod
+    def test_type() -> str:
+        return "MEASUREMENT"
+
     @staticmethod
     def abbreviation():
         return "G"
